File: solutions_1483485_0/C++/smn/QualA.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <climits>
#include <map>
#include <set>
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <algorithm>
#include <functional>
#include <sstream>
#include <stack>
#include <queue>
#include <cstring>
using namespace std;

typedef long long LL;
#define REP(i,e) for (int (i) = 0; (i) < (e); ++(i))
#define foreach(__my_iterator,__my_object) for (typeof((__my_object).begin()) __my_iterator = (__my_object).begin(); __my_iterator!= (__my_object).end(); __my_iterator++)

int main(){
  string v[6] = {"ejp mysljylc kd kxveddknmc re jsicpdrysi",
		 "rbcpc ypc rtcsra dkh wyfrepkym veddknkmkrkcd",
		 "de kr kd eoya kw aej tysr re ujdr lkgc jv",
		 "our language is impossible to understand",
		 "there are twenty six factorial possibilities",
		 "so it is okay if you want to just give up"
  };

  map <char, char> mp;
  mp['y'] = 'a';
  mp['e'] = 'o';
  mp['q'] = 'z';
  mp['z'] = 'q';
  
  REP(i,3) REP(j, v[i].length()) mp[v[i][j]] = v[i+3][j];

  int n;
  string str;
  cin >> n;
  getline(cin, str);
  REP(i, n){
    getline(cin, str);
    REP(j, str.length()) str[j] = mp[str[j]];
    printf("Case #%d: ", i+1);
    cout << str << endl;
  }

  return 0;
}

File: solutions_1483485_0/Python/JustinKang/QualA.py
'''Qualification Round Problem A. Speaking in Tongues'''

import sys


def buildTable( source, target ):
    table = {}
    for i in range( len( source ) ):
        table[ source[ i ] ] = target[ i ]

    alphabet = set( 'abcdefghijklmnopqrstuvwxyz ' )
    sourceSet = set( source )

    missingSource = alphabet - sourceSet
    if missingSource:
        assert len( missingSource ) == 1

        targetSet = set( target )
        missingTarget = alphabet - targetSet

        table[ next( iter( missingSource ) ) ] = next( iter( missingTarget ) )

        assert len( table ) == len( alphabet )

    return table


def main( input ):
    source = 'ejp mysljylc kd kxveddknmc re jsicpdrysi' + \
            'rbcpc ypc rtcsra dkh wyfrepkym veddknkmkrkcd' + \
            'de kr kd eoya kw aej tysr re ujdr lkgc jv' + \
            'qee'
    target = 'our language is impossible to understand' + \
            'there are twenty six factorial possibilities' + \
            'so it is okay if you want to just give up' + \
            'zoo'

    assert len( source ) == len( target )

    table = buildTable( source, target )

    count = int( input.readline() )
    for index in range( 1, count + 1 ):
        source = input.readline().replace( '\n', '' )
        target = source.translate( str.maketrans( table ) )
        print( 'Case #{}: {}'.format( index, target ) )


main( sys.stdin )
